File: apps/api/app/schemas/twilio.py
"""API schemas for organization-level Twilio messaging configuration."""

from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, Field, StringConstraints, field_validator

from app.utils.normalization import normalize_phone

MessagingPurpose = Literal["operational", "promotional"]
TwilioReadinessStatus = Literal[
    "ready",
    "degraded",
    "blocked",
    "not_configured",
    "action_required",
    "unknown",
]
TwilioAccountSid = Annotated[str, StringConstraints(pattern=r"^AC[0-9a-fA-F]{32}$")]
TwilioApiKeySid = Annotated[str, StringConstraints(pattern=r"^SK[0-9a-fA-F]{32}$")]
TwilioMessagingServiceSid = Annotated[str, StringConstraints(pattern=r"^MG[0-9a-fA-F]{32}$")]


class TwilioRouteResponse(BaseModel):
    purpose: MessagingPurpose
    enabled: bool
    messaging_service_sid_masked: str | None
    sender_phone_masked: str | None
    a2p_status: str
    advanced_opt_out_status: str
    consent_management_status: str
    capability_evidence: dict = Field(default_factory=dict)
    webhook_id: str
    inbound_webhook_url: str
    status_callback_url: str


class TwilioSettingsResponse(BaseModel):
    enabled: bool
    account_sid_masked: str | None
    api_key_sid_masked: str | None
    api_secret_configured: bool
    auth_token_configured: bool
    legal_messaging_brand: str | None
    operational_disclosure: str | None
    promotional_disclosure: str | None
    sms_terms_url: str | None
    privacy_policy_url: str | None
    support_contact: str | None
    expected_frequency: str | None
    counsel_approved_at: str | None
    compliance_toolkit_enabled: bool
    twilio_edition: str | None
    baa_verified_at: str | None
    compliance_approved_at: str | None
    phi_enabled: bool
    current_version: int
    routes: dict[MessagingPurpose, TwilioRouteResponse]


class TwilioRouteUpdate(BaseModel):
    enabled: bool | None = None
    messaging_service_sid: TwilioMessagingServiceSid | None = None
    sender_phone_e164: str | None = None
    a2p_status: Literal["unconfigured", "pending", "approved", "rejected"] | None = None
    advanced_opt_out_status: Literal["unconfigured", "enabled", "verified"] | None = None
    consent_management_status: Literal["unknown", "available", "unavailable"] | None = None
    capability_evidence: dict | None = None

    @field_validator("messaging_service_sid", mode="before")
    @classmethod
    def blank_service_sid_clears(cls, value: str | None) -> str | None:
        if isinstance(value, str) and not value.strip():
            return None
        return value

    @field_validator("sender_phone_e164")
    @classmethod
    def normalize_sender_phone(cls, value: str | None) -> str | None:
        if value is None or not value.strip():
            return None
        return normalize_phone(value)


class TwilioSettingsUpdate(BaseModel):
    expected_version: int = Field(ge=1)
    enabled: bool | None = None
    account_sid: TwilioAccountSid | None = None
    api_key_sid: TwilioApiKeySid | None = None
    api_secret: str | None = Field(default=None, max_length=255)
    auth_token: str | None = Field(default=None, max_length=255)
    legal_messaging_brand: str | None = Field(default=None, max_length=160)
    operational_disclosure: str | None = Field(default=None, max_length=4000)
    promotional_disclosure: str | None = Field(default=None, max_length=4000)
    sms_terms_url: str | None = Field(default=None, max_length=1000)
    privacy_policy_url: str | None = Field(default=None, max_length=1000)
    support_contact: str | None = Field(default=None, max_length=255)
    expected_frequency: str | None = Field(default=None, max_length=255)
    counsel_approved_at: datetime | None = None
    compliance_toolkit_enabled: bool | None = None
    twilio_edition: str | None = Field(default=None, max_length=40)
    baa_verified_at: datetime | None = None
    compliance_approved_at: datetime | None = None
    phi_enabled: bool | None = None
    routes: dict[MessagingPurpose, TwilioRouteUpdate] | None = None

    @field_validator("account_sid", "api_key_sid", mode="before")
    @classmethod
    def blank_sid_clears(cls, value: str | None) -> str | None:
        if isinstance(value, str) and not value.strip():
            return None
        return value


class TwilioSettingsTestResponse(BaseModel):
    valid: bool
    account_status: str | None
    twilio_edition: str | None
    capabilities: dict[str, bool]
    error: str | None
    warning: str | None


class TwilioSettingsTestRoute(BaseModel):
    messaging_service_sid: TwilioMessagingServiceSid | None = None

    @field_validator("messaging_service_sid", mode="before")
    @classmethod
    def blank_service_sid_clears(cls, value: str | None) -> str | None:
        if isinstance(value, str) and not value.strip():
            return None
        return value


class TwilioSettingsTestRequest(BaseModel):
    account_sid: TwilioAccountSid | None = None
    api_key_sid: TwilioApiKeySid | None = None
    api_secret: str | None = Field(default=None, max_length=255)
    auth_token: str | None = Field(default=None, max_length=255)
    routes: dict[MessagingPurpose, TwilioSettingsTestRoute] | None = None


class TwilioWebhookRotateRequest(BaseModel):
    purpose: MessagingPurpose
    expected_version: int = Field(ge=1)


class TwilioProviderCapabilities(BaseModel):
    send_sms: bool = False
    send_mms: bool = False
    receive_sms: bool = False
    receive_mms: bool = False
    status_callbacks: bool = False


class TwilioRouteReadiness(BaseModel):
    status: TwilioReadinessStatus
    can_send_sms: bool
    can_send_mms: bool
    can_receive: bool
    issues: list[str] = Field(default_factory=list)


class TwilioProviderReadiness(BaseModel):
    status: TwilioReadinessStatus
    credentials_valid: bool
    account_status: str | None
    checked_at: str | None
    capabilities: TwilioProviderCapabilities
    routes: dict[MessagingPurpose, TwilioRouteReadiness]


class TwilioQueueReadiness(BaseModel):
    status: TwilioReadinessStatus
    queued_count: int
    processing_count: int
    failed_count: int
    oldest_queued_at: str | None


class TwilioReconciliationReadiness(BaseModel):
    status: TwilioReadinessStatus
    action_required_count: int
    unresolved_event_count: int
    last_reconciled_at: str | None


class TwilioLocalReadiness(BaseModel):
    queue: TwilioQueueReadiness
    reconciliation: TwilioReconciliationReadiness


class TwilioReadinessIssue(BaseModel):
    code: str
    severity: Literal["info", "warning", "error"]
    message: str
    route: MessagingPurpose | None


class TwilioReadinessResponse(BaseModel):
    overall_status: TwilioReadinessStatus
    checked_at: str | None
    provider: TwilioProviderReadiness
    local: TwilioLocalReadiness
    issues: list[TwilioReadinessIssue]


class TwilioReadinessCheckResponse(BaseModel):
    check_status: Literal["queued", "running"]
    queued_at: str
    readiness: TwilioReadinessResponse

File: apps/web/lib/api/twilio.ts
/**
 * Organization-scoped Twilio messaging configuration and readiness contracts.
 *
 * Read models expose only masked identifiers and configured flags. Full
 * identifiers and secrets are accepted only by write/test requests.
 */

import api from "../api"

export type TwilioMessagingPurpose = "operational" | "promotional"
export type TwilioReadinessStatus =
    | "ready"
    | "degraded"
    | "blocked"
    | "not_configured"
    | "action_required"
    | "unknown"

export interface TwilioRouteSettings {
    purpose: TwilioMessagingPurpose
    messaging_service_sid_masked: string | null
    sender_phone_masked: string | null
    a2p_status: string
    advanced_opt_out_status: string
    consent_management_status: string
    capability_evidence: Record<string, unknown> | null
    inbound_webhook_url: string
    status_callback_url: string
    webhook_id: string
    enabled: boolean
}

export interface TwilioSettings {
    enabled: boolean
    account_sid_masked: string | null
    api_key_sid_masked: string | null
    api_secret_configured: boolean
    auth_token_configured: boolean
    legal_messaging_brand: string | null
    operational_disclosure: string | null
    promotional_disclosure: string | null
    sms_terms_url: string | null
    privacy_policy_url: string | null
    support_contact: string | null
    expected_frequency: string | null
    counsel_approved_at: string | null
    compliance_toolkit_enabled: boolean
    twilio_edition: string | null
    baa_verified_at: string | null
    compliance_approved_at: string | null
    phi_enabled: boolean
    current_version: number
    routes: Record<TwilioMessagingPurpose, TwilioRouteSettings>
}

export interface TwilioRouteSettingsUpdate {
    messaging_service_sid?: string
    sender_phone_e164?: string
    enabled?: boolean
}

export interface TwilioSettingsUpdate {
    enabled?: boolean
    account_sid?: string
    api_key_sid?: string
    api_secret?: string
    auth_token?: string
    legal_messaging_brand?: string | null
    operational_disclosure?: string | null
    promotional_disclosure?: string | null
    sms_terms_url?: string | null
    privacy_policy_url?: string | null
    support_contact?: string | null
    expected_frequency?: string | null
    counsel_approved_at?: string | null
    compliance_toolkit_enabled?: boolean
    twilio_edition?: string | null
    baa_verified_at?: string | null
    compliance_approved_at?: string | null
    phi_enabled?: boolean
    routes?: Partial<Record<TwilioMessagingPurpose, TwilioRouteSettingsUpdate>>
    expected_version: number
}

export interface TwilioCredentialTestRequest {
    account_sid?: string
    api_key_sid?: string
    api_secret?: string
    auth_token?: string
    routes?: Partial<
        Record<
            TwilioMessagingPurpose,
            { messaging_service_sid?: string; sender_phone_e164?: string }
        >
    >
}

export interface TwilioCredentialTestResponse {
    valid: boolean
    account_status: string | null
    twilio_edition: string | null
    capabilities: Record<string, boolean>
    route_capabilities: Partial<
        Record<TwilioMessagingPurpose, Record<string, boolean | string | null>>
    >
    error: string | null
    warning: string | null
}

export interface TwilioProviderCapabilities {
    send_sms: boolean
    send_mms: boolean
    receive_sms: boolean
    receive_mms: boolean
    status_callbacks: boolean
}

export interface TwilioRouteReadiness {
    status: TwilioReadinessStatus
    can_send_sms: boolean
    can_send_mms: boolean
    can_receive: boolean
    issues: string[]
}

export interface TwilioProviderReadiness {
    status: TwilioReadinessStatus
    credentials_valid: boolean
    account_status: string | null
    checked_at: string | null
    capabilities: TwilioProviderCapabilities
    routes: Record<TwilioMessagingPurpose, TwilioRouteReadiness>
}

export interface TwilioQueueReadiness {
    status: TwilioReadinessStatus
    queued_count: number
    processing_count: number
    failed_count: number
    oldest_queued_at: string | null
}

export interface TwilioReconciliationReadiness {
    status: TwilioReadinessStatus
    action_required_count: number
    unresolved_event_count: number
    last_reconciled_at: string | null
}

export interface TwilioReadinessIssue {
    code: string
    severity: "info" | "warning" | "error"
    message: string
    route: TwilioMessagingPurpose | null
}

export interface TwilioReadiness {
    overall_status: TwilioReadinessStatus
    checked_at: string | null
    provider: TwilioProviderReadiness
    local: {
        queue: TwilioQueueReadiness
        reconciliation: TwilioReconciliationReadiness
    }
    issues: TwilioReadinessIssue[]
}

export interface MessagingTemplateVersion {
    id: string
    template_key: string
    version: number
    name: string
    purpose: TwilioMessagingPurpose
    body: string
    status: "draft" | "published" | "retired"
    is_enrollment_confirmation: boolean
    content_classification: "no_phi" | "phi"
}

export function getTwilioSettings(): Promise<TwilioSettings> {
    return api.get<TwilioSettings>("/twilio/settings")
}

export function updateTwilioSettings(
    update: TwilioSettingsUpdate,
): Promise<TwilioSettings> {
    return api.patch<TwilioSettings>("/twilio/settings", update)
}

export function testTwilioCredentials(
    request: TwilioCredentialTestRequest,
): Promise<TwilioCredentialTestResponse> {
    return api.post<TwilioCredentialTestResponse>("/twilio/settings/test", request)
}

export function getTwilioReadiness(): Promise<TwilioReadiness> {
    return api.get<TwilioReadiness>("/twilio/readiness")
}

export function listMessagingTemplates(params?: {
    purpose?: TwilioMessagingPurpose
    status?: "draft" | "published" | "retired"
}): Promise<MessagingTemplateVersion[]> {
    const query = new URLSearchParams()
    if (params?.purpose) query.set("purpose", params.purpose)
    if (params?.status) query.set("status", params.status)
    const suffix = query.size ? `?${query.toString()}` : ""
    return api.get<MessagingTemplateVersion[]>(`/messaging/templates${suffix}`)
}
